server/src/usingDB/controllers/MessageController.js

```diff
@@ -122,30 +141,61 @@ class MessageController {
     const { rows } = await db.query(messages, [req.decodedMessage.id]);
     output = rows;
     try {
-      if (!output) {
-        return res.status(400).send({ message: 'you have not sent any messages' });
+      if (output.length <= 0) {
+        return res.status(404).send({ message: 'you have not sent any messages' });
       }
-      return res.status(200).send(output);
+      return res.status(200).send({
+        status: 'success',
+        data: output,
+      });
     } catch (e) {
-      return res.status(400).send({ message: 'something is wrong with your request' });
+      return res.status(500).send({ message: 'something is wrong with your request' });
     }
   }
 
   static async deleteAMessage(req, res) {
-    // let output = [];
+    const validNumber = req.params.id;
+    const paramValue = /^\+?(0|[1-9]\d*)$/;
+    const result = paramValue.test(validNumber);
+    if (!result) {
+      return res.status(400).send('you have not inputed a valid ID');
+    }
     const messages = 'UPDATE messages SET is_deleted=$1 WHERE reciever=$2 AND id=$3 returning *';
 
     try {
       const { rows } = await db.query(messages, ['true', req.decodedMessage.id, req.params.id]);
       if (!rows[0]) {
-        return res.status(404).send({ message: 'you cannot delete this message' });
+        return res.status(403).send({ message: 'you cannot delete this message' });
       }
       return res.status(200).send({ message: 'the message has been deleted' });
-      // if(!output) {
-      //   return res.status(400).send({'message': 'email does not exist'});
-      // }
     } catch (e) {
-      return res.status(400).send({ message: 'email does not exist' });
+      return res.status(500).send({ message: 'something is wrong with your request' });
+    }
+  }
+
+  static async retractEmail(req, res) {
+    let messageOutput = [];
+    const validNumber = req.params.id;
+    const paramValue = /^\+?(0|[1-9]\d*)$/;
+    const result = paramValue.test(validNumber);
+    if (!result) {
+      return res.status(400).send('you have not entered a valid id');
+    }
+    const getMessage = 'SELECT * FROM messages WHERE id=$1';
+    const { rows } = await db.query(getMessage, [validNumber]);
+    messageOutput = rows[0];
+    if (!messageOutput) {
+      return res.status(404).send({ message: 'message doesnt exist' });
+    }
+    if (messageOutput.status === 'read') {
+      const deleteMessage = `UPDATE messages SET is_deleted = ${'true'} WHERE sender=$1 and id=$2`;
+      const { rows } = db.query(deleteMessage, [req.decodedMessage.id, validNumber]);
+      return res.status(200).send({ message: 'message retracted' });
```

Comment:
include in the message that this operation was not successful because the receiver has read the message.